### ISSUE TITLE
Add data structure to store locale on posts and terms

### ISSUE THREAD

**Issue by simonwheatley** [simonwheatley/wp-post-term-locale#1]
What do we need to bear in mind with storing this data? Off the top of my head:
- An additional column `post_locale` and `term_locale` on `wp_posts` and `wp_terms`
- An index on the locale columns… just a standard `INDEX`/`KEY`?
- The locale column should support [at least 35 characters](https://core.trac.wordpress.org/ticket/28303#comment:18)… `varchar(100)`?

**Comment by simonwheatley** [comment]
The locale data should be optional when inserting data. If no locale is specified, we should assume that post or term is in the locale that the site is currently using.

**Comment by fgirardey** [comment]
Are we forced to prefix this `locale` column ?

**Comment by simonwheatley** [comment]
We just spent some time in Slack [debating](https://wordpress.slack.com/archives/core-i18n/p1449053704000301) how we could or should store the data for post and term locales. In summary: 

**Option 1:** We add a locale column to the `wp_posts` and `wp_terms` tables, each post or term would then have a locale code associated with it.

**Option 2:** We add a locale table, `wp_locale`, with `locale_ID` and `locale_code` columns, and add a `locale_ID` column to `wp_posts` and `wp_terms`. 

Separately: as we're currently talking about locales for posts and terms, we cannot envisage a scenario where we would need special locales to cover "no locale" or "unknown locale".

**Comment by simonwheatley** [comment]
Personally, I'm leaning towards Option 2 above. I think there's actually a precedent for this in `wp_terms`, in that we don't add the terms in a column on `wp_posts` but we join a separate table.

> Are we forced to prefix this locale column ?

@fgirardey I think we should follow WordPress convention where possible, however column naming convention is not consistent on the [`wp_posts` table](https://github.com/WordPress/WordPress/blob/master/wp-admin/includes/schema.php#L162-L191), so I'm open to suggestion :smiley:

**Comment by tfrommen** [comment]
The difference between terms and locale is, however, that a post can have none, one or more terms, while it (most probably) only has one locale, if any.

Anyway, I also would prefer using a separate table for storing the locales.

**Comment by grappler** [comment]
One example of a CPT not needing a locale is from EDD. I have a download which does not need a locale as it the data is same across locales e.g version number and download link. This made it easier for to track sales better with only a single download id.

**Comment by andreasciamanna** [comment]
As said in the chat, I'm not completely against the idea of using a table.

But I'm still not convinced that using an auto-increment ID to identify the content language, when the language already has his own unique ID, is a healthy idea.

**TL;DR;**

The main problem I see of using multiple tables is that you could lose rows and then have some inconsistent state.

**Long version**

I'll try to explain the reason once again, possibly in a better way :)

We have these languages in the Languages table:

| locale_ID | locale_code |
| --- | --- |
| 1 | en |
| 2 | fr |
| 3 | it |

And these posts:

| ID | post_title | locale_ID |
| --- | --- | --- |
| 100 | Post EN | 1 |
| 101 | Post FR | 2 |
| 102 | Post IT | 3 |

After some time, I decide that I don't need the French language anymore, but I want to keep the existing French post (101) for history reasons.

It doesn't really matter in this scenario how WP_Query behaves for this orphan post.

The languages table changes to that:

| locale_ID | locale_code |
| --- | --- |
| 1 | en |
| 3 | it |

Meanwhile my client ask me to add German to the site's language, so the updated languages table is now:

| locale_ID | locale_code |
| --- | --- |
| 1 | en |
| 3 | it |
| 4 | de |

The client sees that after all, there are enough French users and that it might be worth to restore the French language. So I add it back:

| locale_ID | locale_code |
| --- | --- |
| 1 | en |
| 3 | it |
| 4 | de |
| 5 | fr |

Now, the French post 101 is still orphan.
This requires some code to update all posts with `locale_id=2` with the new value.
Which is fine, I guess, but if I could have used just the language code instead, I wouldn't have needed to take any additional action: the former French post would not be former at all.

I know, in terms of performances, using a numeric ID is better than using some varchar(100), but from a practical use, toe it doesn't look the right approach.

I guess that if we could keep the numeric ID, but not make it an auto increment (let's auto increment it "manually"), we could solve the above problem, as long as the ID is not reused by another language, of course.

**Comment by lloc** [comment]
I think that using a locale column in `wp_posts` and `wp_terms` with a numeric key and a separate language table is the right direction. 

It's not _only_ for performance reasons but this could be big deal if you just think of language codes like the one for the formal de_DE translation for example. Without such a table what would we save in a locale column? The complete string? What about custom/language translation codes that we want to support?

I don't agree that removing a language should lead to remove the corresponding row from the table. Instead I would update a column status to mark that language as disabled. So the table should have at least a layout like this:

| locale_ID | locale_code | locale_status |
| --- | --- | --- |
| 1 | en | 1 |
| 2 | fr | 0 |
| 3 | it | 1 |
| 4 | de | 1 |

This is like the status column in `wp_users` or we think about additional columns like in `wp_blogs`.

**Comment by andreasciamanna** [comment]
Sure, this makes sense, as long as we block from removing records on this table through the API.

Provided that I think this would be an acceptable option, WP already have statuses and post types handled as "attributes" of the post record: these are varchars (20) indexed (well, at least as type + status + date + ID).

I guess that querying by post status and post type is something that happens quite often: is that something wrong in WP, or maybe we didn't consider some other disadvantage of using a dedicated table?

**Comment by markhowellsmead** [comment]
I feel that the solution using a second table is more practical and more future-friendly.

In other CMS, like TYPO3, the language (referred to here as the locale) is stored in a table and then linked across using the ID. So a page or a piece of content is stored with a reference to language ID 1. In the configuration of the site, the ID 1 is then assigned to e.g. German (de_DE) or Swiss German (de_CH).

It is technically possible to re-assign ID 1 to a different language once the content has already been created, but this is not done in practice and should be clear to a developer, that once a language ID is matched with a locale, then it shouldn't be re-assigned. Or the locale will no longer match the actual content.

The full locale is very important, as there are, for example, several German or French languages, each with different locale requirements, and all of which can run concurrently in the same installation.

Performance shouldn't be an issue if the database is correctly indexed, as terms already work in this way.

@sciamannikoo Editing the language table at a later date is a dangerous operation, but it shouldn't be technically prohibited. Auto-incrementing saves a lot of work. A locale should be deactivated when it's no longer in use, not removed. Otherwise, there is no chance of restoring the former content cleanly. Reversible deactivation instead of deletion is standard procedure.

**Comment by barryceelen** [comment]
I’m probably stating the obvious and/or repeating what @mhmli wrote, but I thought it important to point out the fact that ‘locale’ and ‘language’ are different things. A locale encompasses more than just the language, it also defines a default time and date format and for instance, although currently not used in WordPress core, number and currency formats.

A post written in a specific language could be used in multiple locale versions of a site. One might want to reuse a product post type written in German for the localized version of your site aimed at for instance Germany, Austria and Switzerland, where currency or date format might be different but language (mostly) the same.

It is necessary to be able to assign a single post or term to multiple locales. This would rule out using a locale column in wp_posts and wp_terms with a numeric key and a separate language table as @lloc  suggested.

I think the problem of losing context when deleting a locale should best be handled by the author of the plugin which implements the structure proposed in this proof of concept.

**Comment by markhowellsmead** [comment]
The data needs to be cleanly stored and assigned to the specific language/locale combination for which it is intended. Storing it as just `de` is impractical for sites which target content to different regions, where language syntax can differ considerably. For example, the difference between German and Swiss German is huge.

If a site is available to Germany, Switzerland and Austria, where missing content for CH or AT “falls back” to `de_DE`, this needs to be handled by the theme or plugin.

I would strongly advise that a single database column contain this language/locale combination for each piece of localized content.

**Comment by tfrommen** [comment]
Re @mhmli:

> I would strongly advise that a single database column contain this language/locale combination for each piece of localized content.

So you want to have a `locale` column in both `wp_posts` and `wp_terms` table, right?

I would strongly advise to use a separate table instead, and reference it by a unique field (for instance, an auto-incremented ID). This way, plugin authors can easily extend the table as there is [lots of different data to attach to a locale](https://github.com/inpsyde/multilingual-press/blob/5397b119b13fae79cc19b5d07aba8dae5ceb40f5/src/inc/db/Mlp_Db_Languages_Schema.php#L49-L61).

**Comment by markhowellsmead** [comment]
Aopologies, I contradicted myself. Yes, there should be a reference column, which points to an extensible locales table.
